README.md
# FlipDotsPi

WIP. A RaspberryPi server (in Julia) for an [AlphaZeta flip dot](https://flipdots.com/en/products-services/flip-dot-boards-xy5/) display.

_Current:_ Scroll a single message on a flip dot board.
_Future:_ Set up a server on the pi to handle incoming messages.

## Set up

These instructions are for a Raspberry Pi Zero for a 7x28 AlphaZeta display; ymmv on any other controller with any other display.

### Disable auto-running Maestro service

If you're starting from a Raspberry Pi Zero that has been provided by/preconfigured by AlphaZeta, with their Maestro software installed, this section is for you.  If you're starting from your own Raspberry Pi Zero, skip to the following section.

1. Following the instructions provided with the board, set your pi up to connect to your local network:
    1. Plug in/turn on your Pi+display
    2. From your computer, connect to the Maestro wifi network
    3. In your browser, go to `http://flipdots.local/`. If you haven't yet, play around and make sure that you're able to send commands and that your board displays them.
    3. In the  the `http://flipdots.local/admin` panel, update SSID/password to be your local wifi network.
    4. Connect your computer to that same local wifi network.

2. SSH onto the Pi: `ssh pi@flipdots.local` (default pw unless/until you change it: `flipdots`)

3. (Pedantic) Change time via config: `sudo raspi-config`

4. Double-check to make sure your wifi network will be autoconnected next time the Pi starts up. Do `cat /etc/wpa_supplicant/wpa_supplicant.conf`; you should see your wifi network as an entry of the form:
```
network={
  ssid="<SSID>"
  psk="<password>"
  priority=1
}
```
If not, edit this file to add it.

5. Disable Maestro from running on Pi startup:
```
sudo rm /etc/systemd/system/maestro.service
sudo systemctl daemon-reload
```
Then reboot for the changes to take effect:
```
sudo reboot
```

### Install Julia on Raspberry Pi Zero
With many thanks to [@terasakisatoshi](https://discourse.julialang.org/u/terasakisatoshi), via [these instructions](https://discourse.julialang.org/t/have-a-try-julia-v1-5-1-for-arm32bit/45558),:

1. On the Pi, do
```
curl -fsSL get.docker.com -o get-docker.sh && sh get-docker.sh
sudo gpasswd -a $USER docker
sudo reboot
JL_VERSION=v1.5.3
IMAGE_NAME=terasakisatoshi/jlcross:rpizero-${JL_VERSION}
CONTAINER_NAME=jltmp_${JL_VERSION}
docker run --name ${CONTAINER_NAME} $IMAGE_NAME /bin/bash
docker cp ${CONTAINER_NAME}:/home/pi/julia-${JL_VERSION} .
docker rm ${CONTAINER_NAME}
sudo ln -s /home/pi/julia-1.5.3/bin/julia /usr/bin/julia
```

3. Sanity check:
```
which julia
julia --version
```
should yield ` ` and ` `, respectively

4. When running Julia for the first time, you'll need to install the package manager manually (Why? Not sure! It failed otherwise.)
```
rm -rf ~/.julia/registries/General
julia   # launch the REPL
```
Then from the REPL do
```julia
using Pkg
Pkg.Registry.add("General")
```

### Build and install libserialport

To send messages to the display from Julia, we'll use [LibSerialPort.jl](https://github.com/JuliaIO/LibSerialPort.jl). This library typically depends on `libserialport` binaries (via [`libserialport_jll.jl`](https://github.com/JuliaBinaryWrappers/libserialport_jll.jl)), but `libserialport_jll.jl` doesn't automatically build artifactsfor the Pi Zero architecture, so we'll need to build them ourselves and then move them to the expected location.

1. Install the `LibSerialPort.jl` dependency:
```
julia -e "using Pkg; Pkg.add("LibSerialPort")"
```

2. Build `libserialport` in a temporary directory:
```
sudo apt-get install autoconf
sudo apt-get install libtool
sudo apt install git

git clone https://github.com/sigrokproject/libserialport.git
cd libserialport
./autogen.sh
./configure
make
sudo make install
```

3. Confirm that the binaries have been built/installed correctly:
```
ls /usr/local/lib/
```
should yield at least
```
libserialport.a   libserialport.so    libserialport.so.0.1.0
libserialport.la  libserialport.so.0
```

4. Copy the binaries into the directory LibSerialPort.jl will look for it in:
```
sudo cp /usr/local/lib/libserialport* ~/.julia/artifacts/3b640a4f5595d49c1a13088bb157b3098a0c3d82/lib
```

5. Enable the Pi to write to the serial ports (then reboot for it to take effect):
```
sudo usermod -a -G dialout pi
sudo reboot
```

5. Try it! From Julia, do:
```julia
using LibSerialPort
srl = LibSerialPort.open("/dev/ttyS0", 57600; mode=SP_MODE_WRITE)
```
...this doesn't do anything yet, but it also shouldn't error.

### Additional Pi configuration

If you want to use do any additional development, including cloning this repo, you may want to set up some combination of git and tmux. If not, you can skip this section.

1. (Optional) Install and set up tmux:
```
sudo apt-get update
sudo apt-get install tmux -y
touch ~/.tmux.conf
echo "set -g mouse on" >> ~/.tmux.conf
```

2. Get github credentials. For the account you want to have read (or read/write access) on the pi, generate a Personal Access Token (PAT): on Github, do `Settings > Developer Settings > Personal Access Tokens` and generate a token. Don't navigate away from this screen yet---you'll need access to the generated token in a later step.

3. Clone this repo
```
git config credential.helper store
git clone https://github.com/hannahilea/FlipDotsPi.git
```
When prompted for a username, give your github username; when prompted for a password, use the PAT you generated in step 1.

### Displaying messages

Currently there are two options for displaying messages with this package: via command line and via REPL.

To display a one-time message on the Pi, use the command line:
```
cd FlipDotsPi
julia --project=. fliptdot_message.jl "FLIP FLIP HOORAY!!!"
```
This will then display the message `FLIP FLIP HOORAY!!!`. (Note: it may take some time to precompile and display the message; don't be discouraged!) The quotes around the message are required.

If you don't see the message, you may need to adjust your port and/or baudrate, which can be done with flags. To see these configuration options, do
```
julia --project=. scroll_message.jl --help
```

Alternatively, if you want to play around with displaying multiple messages and don't want to deal with the precompilation overhead for each message, run the code via the REPL: `julia --project=.`

Then do:
```julia
include("scroll_message.jl")
scroll_message("HUZZAH!")
scroll_message("O_o")
scroll_message("YAY!")
```

### FAQ

_Q. Even though I disabled Maestro from automatically running, I'd like to manually start it. How do I do that_

A. Do
```
maestro/start.sh
```
then connect by loading [`flipdots.local/`](http://flipdots.local) in the browser.


_Q. How do I shut the pi down when I'm done using it?_

A. It is recommended that the AlphaZeta boards be set to all white before longer term storage, before shutting down the board. Once you've done that, do `sudo shutdown -h now`.


_Q. I can run messages on the display with Maestro, but something about my Julia installation failed. Is there a faster way to bisect the error?_

A. You can try installing/running [a different flipdots scroller](https://github.com/vwyf/vwyf_door_sensor/tree/2a281e274d4d14e7d020158d55fd5dc94bfccd13) via Python. (Be sure to first follow the above steps to stop Maestro.)
1. Set up:
```
sudo apt-get install python-pip
pip install pyserial
cd
wget https://raw.githubusercontent.com/vwyf/vwyf_door_sensor/2a281e274d4d14e7d020158d55fd5dc94bfccd13/flipdot_demon.py
python flipdot_demon.py
```

2. Edit the script (`nano flipdot_demon.py`) to change...
...line 135 from
```
with serial.Serial("/dev/ttyUSB0", 9600) as srl:
```
to
```
with serial.Serial("/dev/serial0", 57600) as srl:
```
and also line 112 from
```
panel_address,
```
to
```
0xFF,
```
At the same time, you can also edit the message to be whatever you'd like by editing the `question_string`.

3. Run it: `python flipdot_demon.py`

### Acknowledgments

Thanks to the follow resources for helping get us up and running:
- [Installing Julia (plus additional required binaries) on the Pi Zero](https://discourse.julialang.org/t/have-a-try-julia-v1-5-1-for-arm32bit/45558/22)
- [Understanding how messages are serialized/constructed](https://ksawerykomputery.pl/tools/flipdigits-player), as well as general inspiration.
- [Example of scrolling message via Python](https://github.com/vwyf/vwyf_door_sensor/tree/2a281e274d4d14e7d020158d55fd5dc94bfccd13)


changelog.md
## TODO

- [ ] Update readme for now-working changelog instructions!
  - remove installation cruft from old attempts

- sysimage

- argparsing for script

- set up server for pi
  - install julia on pi
  - run same startup steps as for previous pi
- auto-run it on boot


## Work log

### Jan 13 2022
Trying to build libserialport manually (instructions here: https://github.com/sigrokproject/libserialport):

1. remove old one:
```
sudo apt-get remomve libserialport-dev
sudo apt-get remomve libserialport0
```
verify nothing:
```
apt list --installed | grep libserialport
```

2. build new one
```
sudo apt-get install autoconf
sudo apt-get install libtool

git clone https://github.com/sigrokproject/libserialport.git
./autogen.sh
./configure
make
sudo make install
```
confirm existance:
```
ls /usr/local/lib/
```

3. Copy it over:
```
sudo cp /usr/local/lib/libserialport* ~/.julia/artifacts/3b640a4f5595d49c1a13088bb157b3098a0c3d82/lib
```
confirm:
```
ls ~/.julia/artifacts/3b640a4f5595d49c1a13088bb157b3098a0c3d82/lib
```
4. Try it:
```julia
using LibSerialPort
srl = LibSerialPort.open("/dev/ttyS0", 57600; mode=SP_MODE_WRITE)

#note: using /dev/serial0 fails?!? still?!? dunno why.

```

### Jan 9 2022
- Spent a lot of time trying to figure out how to get LibSerialPort deps running. Maybe succeeded? Instructions are now up to date; lots of thanks to https://discourse.julialang.org/t/have-a-try-julia-v1-5-1-for-arm32bit/45558/22

- Looks like our serial port isn't enabled by default; must have been previously set up by maestro? Trying to figure out how to set that up.
    - For future: in maestro/displayConfig.yml, option to uncomment type to make it different type of screen

- After a bunch of mucking around, unable to make LibSerialPorts.jl work, reran the old python script as a sanity-check. It worked. -__- Giving up on LibSerialPorts, moving to SerialPorts.jl (which wraps the python library).
- For future reference, to get the

### Jan 8 2022

- made a repo! (https://github.com/hannahilea/FlipDotsPi)

- Poked around at install scripts:
    - `cat maestro/install_maestro.sh`
    - `cat maestro/install_autohotspot.sh`
    - We probably want to disable the autohotspot? At least after we set up the wifi correctly...Actually it seems like maybe not, that's a good fallback if for whatever reason our private network is down (like it was when we started today).

- Time to disable the run-on-start maestro! How do we do that?
    - Thanks to the helpful https://www.thedigitalpictureframe.com/ultimate-guide-systemd-autostart-scripts-raspberry-pi/

- Spent a lot of time trying to figure out how to get LibSerialPort deps running. Did not succeed. Instructions are now up to date; lots of thanks to https://discourse.julialang.org/t/have-a-try-julia-v1-5-1-for-arm32bit/45558/22

### Jan 4 2022

From maestro, connected to our wifi.
SSH'd in, updated local time on pi.
Ran python script:

1. Copy their script to the pi
```
cd
mkdir mucking
cd mucking
wget https://raw.githubusercontent.com/vwyf/vwyf_door_sensor/2a281e274d4d14e7d020158d55fd5dc94bfccd13/flipdot_demon.py
```

2. Edit that script
`nano flipdot_demon.py`

to change...
...line 135 from
```
with serial.Serial("/dev/ttyUSB0", 9600) as srl:
```
to
```
with serial.Serial("/dev/serial0", 57600) as srl:
```
and also line 112 from
```
panel_address,
```
to
```
0xFF,
```
At the same time, you can also edit the message to be whatever you'd like by editing the `question_string`.


2. Install pip+deps, woo
Not using virtualenv or python3 b/c we won't be using python for long anyway.
```
sudo apt-get install python-pip
pip install pyserial
```

3. Figure out how to stop maestro (which runs by default on the pi)
Probably a better way to do this, but we did `htop` and then killed all the maestro-looking proccesses (which were running as `root`).

3. Run it!
```
python flipdot_demon.py
```

(Thanks to https://ksawerykomputery.pl/tools/flipdigits-player for additional info that made these tweaks understandable!)

4. Be fully done:
```
sudo shutdown -h now
```

Next up:
- no moar python

Other cool projects with flipdots:
- https://hackaday.com/2021/11/09/flip-dot-oscilloscope-is-flippin-awesome/



scroll_message.jl
# Rewrite of https://raw.githubusercontent.com/vwyf/vwyf_door_sensor/2a281e274d4d14e7d020158d55fd5dc94bfccd13/flipdot_demon.py

### flipdot display
###
### 0:1200
### 1:2400
### 2:4800
### 3:9600 <--- this should be set, means 1-ON 2-ON 3-OFF
### 4:19200
### 5:38200 <-- do not use, most probably wrong speed programmed
### 6: 9600
### 7: 9600
### 8: 9600
###
### 0x80 beginning
###___________________
### 0x81 - 112 bytes / no refresh / C+3E
### 0x82 - refresh
### 0x83 - 28 bytes of data / refresh / 2C
### 0x84 - 28 bytes of data / no refresh / 2C
### 0x85 - 56 bytes of data / refresh / C+E
### 0x86 - 56 bytes of data / no refresh / C+E
### ---------------------------------------
### address or 0xFF for all
### data ... 1 to number of data buytes
### 0x8F end

using LibSerialPort
using ArgParse

const DEFAULT_PANEL_WIDTH = 28

const DEFAULT_PORTNAME = "/dev/ttyS0"  # "/dev/serial0"
const DEFAULT_BAUDRATE = 57600
const DEFAULT_LOOPCOUNT = 1
const DEFAULT_SCROLLPAUSE = 0.3f0
# const DEFAULT_DISPLAYTYPE = "dots"

# font = Dict(
#     ' ' => UInt8.([0]),
#     '!' => UInt8.([46]),
#     '+' => UInt8.([24,  126, 126,  24,  0]),
#     '-' => UInt8.([24,   24,  24,  24,  0]),
#     '0' => UInt8.([62,   65,  65,  62,  0]),
#     '1' => UInt8.([0,   66,  127, 64,  0]),
#     '2' => UInt8.([98,   81,  73,  70,  0]),
#     '3' => UInt8.([34,   65,  73,  54,  0]),
#     '4' => UInt8.([56,   36,  34, 127, 32]),
#     '5' => UInt8.([79,   73,  73,  49,  0]),
#     '6' => UInt8.([62,   73,  73,  50,  0]),
#     '7' => UInt8.([3,    1,   1, 127,  0]),
#     '8' => UInt8.([54,   73,  73,  54,  0]),
#     '9' => UInt8.([38,   73,  73,  62,  0]),
#     'A' => [0x3C, 0x0A, 0x0A, 0x3C],
#     'B' => [0x3E, 0x2A, 0x2A, 0x14],
#     'C' => [0x1C, 0x22, 0x22, 0x14],
#     'D' => [0x3E, 0x22, 0x22, 0x1C],
#     'E' => [0x3E, 0x2A, 0x2A],
#     'F' => [0x3E, 0x0A, 0x0A],
#     'G' => [0x1C, 0x22, 0x2A, 0x2A],
#     'H' => [0x3E, 0x08, 0x08, 0x3E],
#     'I' => [0x3E],
#     'J' => [0x10, 0x20, 0x20, 0x1E],
#     'K' => [0x3E, 0x08, 0x14, 0x22],
#     'L' => [0x3E, 0x20, 0x20],
#     'M' => [0x3E, 0x04, 0x08, 0x04],
#     'N' => [0x3E, 0x04, 0x08, 0x3E],
#     'O' => [0x1C, 0x22, 0x22, 0x1C],
#     'P' => [0x3E, 0x0A, 0x0A, 0x1C, 0x04],
#     'Q' => [0x1C, 0x22, 0x12, 0x2C],
#     'R' => [0x3E, 0x0A, 0x1A, 0x24],
#     'S' => [0x24, 0x2A, 0x2A, 0x12],
#     'T' => [0x02, 0x02, 0x3E, 0x02, 0x02],
#     'U' => [0x1E, 0x20, 0x20, 0x1E],
#     'V' => [0x06, 0x18, 0x20, 0x18, 0x6],
#     'W' => [0x1E, 0x20, 0x1E, 0x20, 0x1E],
#     'X' => [0x36, 0x08, 0x08, 0x36],
#     'Y' => [0x2E, 0x28, 0x28, 0x1E],
#     'Z' => [0x32, 0x2A, 0x2A, 0x26],
# )

# all_bright = [
#     0x80,  # header
#     0x83,  # 28 bytes refresh
#     0xFF,  # address
#     0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F, # 28 bytes data
#     0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F,
#     0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F,
#     0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F, 0x7F,
#     0x8F # EOT
# ]

# all_dark = [
#     0x80,  #header
#     0x83,  # 28 bytes refresh
#     0xFF,  # address
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, # 28 bytes data
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
#     0x8F # EOT
# ]

# hello_world = [
#     0x80,  #header
#     0x83,  # 28 bytes refresh
#     0xFF,  # address
#     digitsfont['1'], digitsfont['2'], digitsfont['8'], digitsfont['8'], # 28 bytes data
#     0x00, 0x00, 0x00, #0x00, 0x00, 0x00, 0x00,
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
#     0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
#     0x8F # EOT
# ]

# digitsfont = Dict(
#     ' ' => seg_to_bits([]),
#     '-' => seg_to_bits([1]),
#     '_' => seg_to_bits([4]),
#     '0' => seg_to_bits([2:7]...),
#     '1' => seg_to_bits([5, 6]),
#     '2' => seg_to_bits([7, 6, 1, 3, 4]),
#     '3' => seg_to_bits([7, 6, 1, 5, 4]),
#     '4' => seg_to_bits([2, 1, 6, 5]),
#     '5' => seg_to_bits([7, 2, 1, 5, 4]),
#     '6' => seg_to_bits([7, 2, 3, 4, 5, 1]),
#     '7' => seg_to_bits([7, 6, 5]),
#     '8' => seg_to_bits([1:7]...),
#     '9' => seg_to_bits([1, 2, 4, 5, 6, 7]),
#     'A' => seg_to_bits([1, 2, 3, 5, 6, 7]),
#     'B' => seg_to_bits([1, 2, 3, 4, 5]),
#     'C' => seg_to_bits([1, 3, 4]),
#     'D' => seg_to_bits([1, 3, 4, 5, 6]),
#     'E' => seg_to_bits([7, 2, 1, 3, 4]),
#     'F' => seg_to_bits([7, 2, 1, 3]),
#     'G' => seg_to_bits([7, 2, 3, 4, 5]),
#     'H' => seg_to_bits([2, 3, 1, 5]),
#     'I' => seg_to_bits([5]),
#     'J' => seg_to_bits([6, 5, 4, 3]),
#     'K' => seg_to_bits([7, 2, 1, 3, 5]),
#     'L' => seg_to_bits([2, 3, 4]),
#     'M' => seg_to_bits([2, 3, 7, 6, 5]),
#     'N' => seg_to_bits([3, 1, 5]),
#     'O' => seg_to_bits([1, 3, 4, 5]),
#     'P' => seg_to_bits([3, 2, 7, 6, 1]),
#     'Q' => seg_to_bits([1, 2, 7, 6, 5]),
#     'R' => seg_to_bits([3, 1]),
#     'S' => seg_to_bits([2, 1, 5, 4]),
#     'T' => seg_to_bits([2, 3, 1, 4]),
#     'U' => seg_to_bits([3, 4, 5]),
#     'V' => seg_to_bits([2, 3, 4, 5, 6]),
#     'W' => seg_to_bits([2, 3, 4, 5, 6, 1]),
#     'X' => seg_to_bits([2, 3, 6, 5, 1]),
#     'Y' => seg_to_bits([2, 1, 6, 5, 4]),
#     'Z' => seg_to_bits([7, 6, 3, 4]),
# )

# function seg_to_bits(i_segs::AbstractVector)
#     bitstr = zeros(Int, 8)
#     foreach(i -> bitstr[i] = 1 , i_segs)
#     return parse(UInt8, reverse(join(bitstr)), base=2)
# end

# function show_slice(t, srl, msg; wrap_msg=false)
#     transmission = [
#         0x80, #header
#         0x83, # 28 bytes, refresh
#         0xFF, # panel address
#     ]
#     for i in t:(t + 27)
#         if i <= length(msg)
#             append!(transmission, msg[i])
#         elseif wrap_msg
#             append!(transmission, msg[i % length(msg)])
#         else
#             append!(transmission, 0x00)
#         end
#     end
#     append!(transmission, 0x8F) # EOT
#     write(srl, transmission)
# end

# function construct_question(message; panel_width=DEFAULT_PANEL_WIDTH, displaytype="dots")
#     # loop over question string and add columns to question
#     font_dict = font
#     add_spacer = true
#     if displaytype != "dots"
#         font_dict = digitsfont
#         add_spacer = false
#     end
#     message = uppercase(message) # TODO: when supporting lowercase on dots board, move into digits only
#     question = UInt8[]
#     for c in message
#         if c in keys(font_dict)
#             append!(question, font_dict[c])
#             add_spacer && append!(question, 0) # put space between letters
#         end
#     end

#     # if question is fewer than 28 columns pad it out with spaces
#     while length(question) < panel_width
#         append!(question, font_dict[' '])
#     end
#     displaytype != "dots" && append!(question, font_dict[' '])
#     return question
# end

# function flash_display(srl, sleep_sec=0.5)
#     write(srl, all_bright)
#     sleep(sleep_sec)
#     write(srl, all_dark)
#     sleep(sleep_sec)
#     return nothing
# end

# function scroll_message(message; loopcount::Int=DEFAULT_LOOPCOUNT, baudrate=DEFAULT_BAUDRATE,
#                         portname=DEFAULT_PORTNAME, scrollpause=DEFAULT_SCROLLPAUSE,
#                         panel_width=DEFAULT_PANEL_WIDTH, displaytype=DEFAULT_DISPLAYTYPE, verbose=false)
#     # question = construct_question(message; panel_width, displaytype)
#     # verbose && println("Msg to display: $question")
#     # LibSerialPort.open(portname, baudrate) do srl
#         verbose && println("Displaying message")

#         loopcount == 0 && show_slice(1, srl, question; wrap_msg=false)

#         for i in 1:loopcount
#             wrap_msg = i != loopcount
#             for t in 1:length(question)
#                 show_slice(t, srl, question; wrap_msg)
#                 sleep(scrollpause)
#             end
#         end
#     end
# end

#####
## IFF running from the REPL
#####

# function parse_commandline()
#     s = ArgParseSettings(; description="Scroll a message on a FlipDots display.")
#     @add_arg_table s begin
#         "--loopcount"
#             help = "Number of times to scroll the message"
#             arg_type = Int
#             default = DEFAULT_LOOPCOUNT
#         "--baudrate"
#             help = "Baudrate of the display"
#             arg_type = Int
#             default = DEFAULT_BAUDRATE
#         "--portname"
#             help = "Serial port name for the display"
#             arg_type = String
#             default = DEFAULT_PORTNAME
#         "--scrollpause"
#             help = "Pause between scroll updates (sec)"
#             arg_type = Float32
#             default = DEFAULT_SCROLLPAUSE
#         "--panelwidth"
#             help = "Width of the display (# dots)"
#             arg_type = Int
#             default = DEFAULT_PANEL_WIDTH
#         "--displaytype"
#             help = "Font option: flipdots display (`dots`) or flipdigits display (`digits`)"
#             arg_type = String
#             default = DEFAULT_displaytype
#         "--verbose"
#             help = "Show additional debug statements"
#             action = :store_true
#         "message"
#             help = "Message to be displayed; supports uppercase letters, numbers, and limited symbols." #TODO list symbols
#             required = true
#     end
#     return parse_args(s)
# end

# # ...when running as script (not from REPL):
# if !isinteractive()
#     d = parse_commandline()
#     args = NamedTuple{Tuple(Symbol.(keys(d)))}(values(d))
#     args.verbose && println("Parsed args: $args")
#     scroll_message(args.message; loopcount=args.loopcount, baudrate=args.baudrate,
#                    portname=args.portname, scrollpause=args.scrollpause, verbose=args.verbose,
#                    panel_width=args.panelwidth, displaytype=args.displaytype)
# end

#####
## Fun snippets
#####

function drumbeat_snippet(pause=.1)
    for i in 1:16
        scroll_message("1 1 1 1 "; loopcount=0, displaytype="digits")
        sleep(pause)
        scroll_message("1 1    1 "; loopcount=0, displaytype="digits")
        sleep(pause)
        scroll_message(" 1     1 "; loopcount=0, displaytype="digits")
        sleep(pause)
        scroll_message("      "; loopcount=0, displaytype="digits")
        sleep(pause)
    end
end

function drumbeat_snippet2(pause=.1)
    for i in 1:16
        scroll_message("1 1 1 1 1 1 1 1 1 1 1 1 1 1"; loopcount=0, displaytype="digits")
        sleep(pause)
        scroll_message("1   1 1 1   1 1 1 1   1 1 1"; loopcount=0, displaytype="digits")
        sleep(pause)
        scroll_message("1   1 1 1             1 1 1"; loopcount=0, displaytype="digits")
        sleep(pause)
        scroll_message(""; loopcount=0, displaytype="digits")
        sleep(pause)
    end
end